Source/utils/pcx_to_cl2.cpp

```diff
@@ -0,0 +1,266 @@
+{
+	const uint8_t *begin = src;
+	const uint8_t *prevColorBegin = src;
+	unsigned prevColorRunLength = 1;
+	uint8_t prevColor = *src++;
+	while (--length > 0) {
+		const uint8_t color = *src;
+		if (prevColor == color) {
+			++prevColorRunLength;
+		} else {
+			// A tunable parameter that decides at which minimum length we encode a fill run.
+			// 3 appears to be optimal for most of our data (much better than 2, rarely very slightly worse than 4).
+			constexpr unsigned MinFillRunLength = 3;
+			if (prevColorRunLength >= MinFillRunLength) {
+				AppendCl2PixelsRun(begin, prevColorBegin - begin, out);
+				AppendCl2FillRun(prevColor, prevColorRunLength, out);
+				begin = src;
+			}
+			prevColorBegin = src;
+			prevColorRunLength = 1;
+			prevColor = color;
+		}
+		++src;
+	}
+	AppendCl2PixelsRun(begin, prevColorBegin - begin, out);
+	AppendCl2FillRun(prevColor, prevColorRunLength, out);
+}
+
+size_t GetReservationSize(size_t pcxSize)
+{
+	// For the most part, CL2 is smaller than PCX, with a few exceptions.
+	switch (pcxSize) {
+	case 2352187: // ui_art\hf_logo1.pcx
+		return 2464867;
+	case 172347: // ui_art\creditsw.pcx
+		return 172347;
+	case 157275: // ui_art\credits.pcx
+		return 173367;
+	default:
+		return pcxSize;
+	}
+}
+
+} // namespace
+
+std::optional<OwnedCelSpriteWithFrameHeight> PcxToCl2(SDL_RWops *handle, int numFramesOrFrameHeight, std::optional<uint8_t> transparentColor, SDL_Color *outPalette)
+{
+	int width;
+	int height;
+	uint8_t bpp;
+	if (!LoadPcxMeta(handle, width, height, bpp)) {
+		SDL_RWclose(handle);
+		return std::nullopt;
+	}
+	assert(bpp == 8);
+
+	unsigned numFrames;
+	unsigned frameHeight;
+	if (numFramesOrFrameHeight > 0) {
+		numFrames = numFramesOrFrameHeight;
+		frameHeight = height / numFrames;
+	} else {
+		frameHeight = -numFramesOrFrameHeight;
+		numFrames = height / frameHeight;
+	}
+
+	ptrdiff_t pixelDataSize = SDL_RWsize(handle);
+	if (pixelDataSize < 0) {
+		SDL_RWclose(handle);
+		return std::nullopt;
+	}
+
+	pixelDataSize -= PcxHeaderSize;
+
+	std::unique_ptr<uint8_t[]> fileBuffer { new uint8_t[pixelDataSize] };
+	if (SDL_RWread(handle, fileBuffer.get(), pixelDataSize, 1) == 0) {
+		SDL_RWclose(handle);
+		return std::nullopt;
+	}
+
+	// CEL header: frame count, frame offset for each frame, file size
+	std::vector<uint8_t> cl2Data;
+	cl2Data.reserve(GetReservationSize(pixelDataSize));
+	cl2Data.resize(4 * (2 + static_cast<size_t>(numFrames)));
+	WriteLE32(cl2Data.data(), numFrames);
+
+	// We process the PCX a whole frame at a time because the lines are reversed in CEL.
+	auto frameBuffer = std::unique_ptr<uint8_t[]>(new uint8_t[static_cast<size_t>(frameHeight) * width]);
+
+	const unsigned srcSkip = width % 2;
+	uint8_t *dataPtr = fileBuffer.get();
+	for (unsigned frame = 1; frame <= numFrames; ++frame) {
+		WriteLE32(&cl2Data[4 * static_cast<size_t>(frame)], static_cast<uint32_t>(cl2Data.size()));
+
+		// Frame header: 5 16-bit offsets to 32-pixel height blocks.

```

Comment:
Regarding converting assets to Cl2 offline. If we on load decode the first block and divide by 32 then we would know the width? Would we need to also decode the full image to know the height?
I think pngs2cel might now encode this, but that can be ignored.

Reply:
> If we on load decode the first block and divide by 32 then we would know the width? 

Yes but the width calculation doesn't work for CL2 files that are fewer than 32 lines tall.

> Would we need to also decode the full image to know the height?

Yes.

Let's keep the more general discussion outside of this PR (e.g. in the issue), unless you think it affects it somehow?